dhall-toml/src/Dhall/Toml.hs

```diff
@@ -7,33 +7,96 @@ module Dhall.Toml
     , dhallToToml
     ) where
 
-
+import Control.Monad     (foldM)
 import Control.Exception (Exception)
+import Data.Text         (unpack)
 import Data.Void         (Void)
-import Dhall.Core        (Expr)
+import Dhall.Core        (Expr, DhallDouble(..))
 import Dhall.Parser      (Src)
 import Toml.Type.TOML    (TOML)
+import Toml.Type.Key     (Piece(Piece), Key(Key))
+import Toml.Type.Printer (pretty)
+
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Data.Text.IO       as Text.IO
+import qualified Dhall.Map          as Map
+import qualified Dhall.Core         as Core
+import qualified Dhall.Import
+import qualified Dhall.Parser
+import qualified Dhall.TypeCheck
+import qualified Toml.Type.TOML     as T
+import qualified Toml.Type.Value    as TVal
 
 
 -- TODO: populate with actual errors
 data CompileError
-    = CompileError String
+    = Unimplemented String
+    | Unsupported (Expr Void Void)
+    | NotARecord (Expr Void Void)
+
 
 instance Show CompileError where
-    show (CompileError s) = "compile error: " ++ s
+    show (Unimplemented s) = "unimplemented: " ++ s
+    show (Unsupported e) = "unsupported: " ++ show e
+    show (NotARecord e) = "The root object converted to TOML must be a record, got " ++ show e
 
 instance Exception CompileError
 
 dhallToToml :: Expr s Void -> Either CompileError TOML
-dhallToToml _ = Left $ CompileError  "not implemented"
+dhallToToml e0 = do
+    let norm = Core.alphaNormalize $ Core.normalize e0
```

Comment:
Like in `dhall-csv`, we probably don't need α-normalization: https://github.com/dhall-lang/dhall-haskell/pull/2226#discussion_r656266693
```suggestion
    let norm = Core.normalize e0
```